{
 "cells": [
  {
   "cell_type": "raw",
   "id": "dc2e2013",
   "metadata": {},
   "source": [
    "---\n",
    "format:\n",
    "  html:\n",
    "    code-block-background: true\n",
    "    code-block-bg: true\n",
    "    code-block-border-left: \"#31BAE9\"\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5042d3f7",
   "metadata": {},
   "source": [
    "# Streamlining the analysis process\n",
    "\n",
    "In the previous chapters, we have seen various steps in generating taxonomic assignment for raw reads data. Manual execution of all those steps, one by one, could be time-consuming and may be prone to errors. To address these challenges, there are several workflow management tools available (e.g., snakemake, nextflow). These tools automate the execution of the entire pipeline and also enable reproducibility of running the same workflow with different datasets.\n",
    "\n",
    "Here, we will see how to use Snakemake, a workflow management tool, to streamline the entire process of taxonomy generation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1177560e",
   "metadata": {},
   "source": [
    "## Snakemake configuration file\n",
    "As the first step, we will create a configuration file where we will keep all needed information for our analysis. It includes project directory, names of output files created in the analysis, parameters values for tools, etc.\n",
    "\n",
    "This configuration file has information in the format of `key:value`. The below file is a configuration file which we are going to use for our short analysis (i.e., primer removal -> asv generation -> taxonomy assignment).\n",
    "\n",
    "```{.yml}\n",
    "## Basic config\n",
    "project: trial\n",
    "scratch: /Users/pankaj/Documents/Metrosert/public_dataset/zakular2014\n",
    "raw_data: partial_raw_data\n",
    "outputDIR: partial_raw_asv\n",
    "manifest: manifest_partial.csv\n",
    "metadata: metadata_partial.tsv\n",
    "\n",
    "## 16S adapters from Zackular et al., 2014\n",
    "primerF: GTGCCAGCMGCCGCGGTAA\n",
    "primerR: GGACTACHVGGGTWTCTAAT\n",
    "primer_err: 0.4\n",
    "primer_overlap: 3\n",
    "\n",
    "## Reference database\n",
    "database: silva-138.1-ssu-nr99-seqs-515f-806r.qza\n",
    "database_classifier: silva_classifier.qza\n",
    "database_tax: silva-138.1-ssu-nr99-tax.qza\n",
    "\n",
    "## DADA2 - ASV flags\n",
    "truncation_err: 2\n",
    "truncation_len-f: 150\n",
    "truncation_len-r: 140\n",
    "truncation_err: 2\n",
    "quality_err: 2\n",
    "\n",
    "## Diversity metrics\n",
    "sampling_depth: 500\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "851b87b6",
   "metadata": {},
   "source": [
    "Now, we will put our analysis codes in one place to be used by Snakemake. To do that, we need to create a file without specifying any extension. For example, we create a file `Snakefile-asv`. Next, we will write our analysis codes in the file a specific format (which we will see later).\n",
    "\n",
    "Basically, we will do the following\n",
    "\n",
    "1. Specify configuration file\n",
    "2. Extract all information provided in configuration file\n",
    "3. Write rules for each analysis step\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65d09967",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Snakemake file - input quality controlled fastq reads to generate asv\n",
    "# Pankaj chejara\n",
    "\n",
    "# Base snakefile: https://github.com/shu251/tagseq-qiime2-snakemake/blob/master/Snakefile-asv\n",
    "\n",
    "configfile: \"config.yaml\"\n",
    "\n",
    "\n",
    "import io\n",
    "import os\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "\n",
    "##########################################################\n",
    "#                 SET CONFIG VARS\n",
    "##########################################################\n",
    "\n",
    "PROJ = config[\"project\"]\n",
    "SCRATCH = config[\"scratch\"]\n",
    "INPUTDIR = config[\"raw_data\"]\n",
    "OUTPUTDIR = config['outputDIR']\n",
    "MANIFEST = config[\"manifest\"]\n",
    "METADATA = config[\"metadata\"]\n",
    "SAMPLING_DEPTH= config['sampling_depth']\n",
    "\n",
    "# Database information\n",
    "DB = config[\"database\"]\n",
    "DB_classifier = config[\"database_classifier\"]\n",
    "DB_tax = config[\"database_tax\"]\n",
    "\n",
    "\n",
    "rule all:\n",
    "  input:\n",
    "    q2_import = SCRATCH + \"/\" + OUTPUTDIR +\"/\" +  PROJ + \"-PE-demux.qza\",\n",
    "    q2_primerRM = SCRATCH + \"/\" + OUTPUTDIR +\"/\" +  PROJ + \"-PE-demux-noprimer.qza\",\n",
    "    # Visualization\n",
    "    raw = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-PE-demux.qzv\",\n",
    "    primer = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-PE-demux-noprimer.qzv\",\n",
    "    # Dada2 results\n",
    "    table = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.qza\",\n",
    "    rep = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-rep-seqs.qza\",\n",
    "    stats = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-stats-dada2.qza\",\n",
    "    stats_viz = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-stats-dada2.qzv\",\n",
    "    # Taxonomic table\n",
    "    sklearn = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" +  PROJ + \"-tax_sklearn.qza\",\n",
    "    table_biom = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"feature-table.biom\",\n",
    "    table_tsv = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.tsv\",\n",
    "    table_tax = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\"  + \"taxonomy.tsv\",\n",
    "    # Phylogenetic outputs\n",
    "    aligned_seqs = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-aligned-rep-seqs.qza\",\n",
    "    aligned_masked = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-masked-aligned-rep-seqs.qza\",\n",
    "    unrooted_tree = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-unrooted-tree.qza\",\n",
    "    rooted_tree = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-rooted-tree.qza\",\n",
    "    # Diversity metrics\n",
    "    output_dir = directory(SCRATCH + \"/qiime2/diversity\")\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 LOAD DATA \n",
    "##########################################################\n",
    "rule import_qiime:\n",
    "  input:\n",
    "    MANIFEST\n",
    "  output:\n",
    "    q2_import = SCRATCH + \"/\" + OUTPUTDIR +\"/\" + PROJ + \"-PE-demux.qza\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_q2.log\"\n",
    "  params:\n",
    "    type=\"SampleData[PairedEndSequencesWithQuality]\",\n",
    "    input_format=\"PairedEndFastqManifestPhred33\"\n",
    "  shell:\n",
    "    \"\"\"qiime tools import \\\n",
    "    --type {params.type} \\\n",
    "    --input-path {input} \\\n",
    "    --output-path {output.q2_import} \\\n",
    "    --input-format {params.input_format} \n",
    "    \"\"\"\n",
    "\n",
    "##########################################################\n",
    "#                 REMOVE PRIMERS\n",
    "##########################################################\n",
    "\n",
    "rule rm_primers:\n",
    "  input:\n",
    "    q2_import = SCRATCH + \"/\" + OUTPUTDIR +\"/\" + PROJ + \"-PE-demux.qza\"\n",
    "  output:\n",
    "    q2_primerRM = SCRATCH + \"/\" + OUTPUTDIR +\"/\" + PROJ +  \"-PE-demux-noprimer.qza\"\n",
    "  log:\n",
    "     SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_primer_q2.log\"\n",
    "\n",
    "  shell:\n",
    "    \"\"\"qiime cutadapt trim-paired \\\n",
    "       --i-demultiplexed-sequences {input.q2_import} \\\n",
    "       --p-front-f {config[primerF]} \\\n",
    "       --p-front-r {config[primerR]} \\\n",
    "       --p-error-rate {config[primer_err]} \\\n",
    "       --p-overlap {config[primer_overlap]} \\\n",
    "       --o-trimmed-sequences {output.q2_primerRM}\"\"\"\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 QC STATS\n",
    "##########################################################\n",
    "\n",
    "rule get_stats:\n",
    "  input:\n",
    "    q2_import = SCRATCH + \"/\" + OUTPUTDIR +\"/\" +  PROJ + \"-PE-demux.qza\",\n",
    "    q2_primerRM = SCRATCH + \"/\" + OUTPUTDIR +\"/\" +  PROJ + \"-PE-demux-noprimer.qza\"\n",
    "  output:\n",
    "    raw = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-PE-demux.qzv\",\n",
    "    primer = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-PE-demux-noprimer.qzv\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ +  \"_getviz_q2.log\"\n",
    "  shell:\n",
    "    \"\"\"\n",
    "     qiime demux summarize --i-data {input.q2_import} --o-visualization {output.raw}\n",
    "     qiime demux summarize --i-data {input.q2_primerRM} --o-visualization {output.primer}\n",
    "    \"\"\"\n",
    "\n",
    "##########################################################\n",
    "#                 DENOISE & ASVs\n",
    "##########################################################\n",
    "\n",
    "rule dada2:\n",
    "  input:\n",
    "    q2_primerRM = SCRATCH + \"/\" + OUTPUTDIR +\"/\" + PROJ + \"-PE-demux-noprimer.qza\"\n",
    "  output:\n",
    "    table = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.qza\",\n",
    "    rep = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-rep-seqs.qza\",\n",
    "    stats = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-stats-dada2.qza\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_dada2_q2.log\"\n",
    "  shell:\n",
    "    \"\"\"qiime dada2 denoise-paired \\\n",
    "        --i-demultiplexed-seqs {input.q2_primerRM} \\\n",
    "        --p-trunc-q {config[truncation_err]} \\\n",
    "        --p-trunc-len-f {config[truncation_len-f]} \\\n",
    "        --p-trunc-len-r {config[truncation_len-r]} \\\n",
    "        --o-table {output.table} \\\n",
    "        --o-representative-sequences {output.rep} \\\n",
    "        --o-denoising-stats {output.stats}\"\"\"\n",
    "\n",
    "\n",
    "rule dada2_stats:\n",
    "  input:\n",
    "    stats = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-stats-dada2.qza\"\n",
    "  output:\n",
    "    stats_viz = SCRATCH + \"/\" + OUTPUTDIR + \"/viz/\" + PROJ + \"-stats-dada2.qzv\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_dada2-stats_q2.log\"\n",
    "  shell:\n",
    "   \"\"\"qiime metadata tabulate \\\n",
    "       --m-input-file {input.stats} \\\n",
    "       --o-visualization {output.stats_viz}\"\"\"\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 TAXONOMIC ASSIGNMENT\n",
    "##########################################################\n",
    "\n",
    "rule assign_tax:\n",
    "  input:\n",
    "    rep = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" +  PROJ + \"-rep-seqs.qza\",\n",
    "    db_classified = DB_classifier\n",
    "  output:\n",
    "    sklearn = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" +  PROJ + \"-tax_sklearn.qza\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ +  \"_sklearn_q2.log\"\n",
    "  shell:\n",
    "    \"\"\"qiime feature-classifier classify-sklearn \\\n",
    "\t  --i-classifier {input.db_classified} \\\n",
    "\t  --i-reads {input.rep} \\\n",
    "\t  --o-classification {output.sklearn}\"\"\"\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 TAXONOMIC TABLE GENERATION\n",
    "##########################################################\n",
    "rule gen_table:\n",
    "  input:\n",
    "    table = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.qza\"\n",
    "  output:\n",
    "    table_biom = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"feature-table.biom\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_exportBIOM_q2.log\"\n",
    "  params:\n",
    "    directory(SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\")\n",
    "  shell:\n",
    "    \"qiime tools export --input-path {input.table} --output-path {params}\"\n",
    "\n",
    "rule convert:\n",
    "  input:\n",
    "    table_biom = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"feature-table.biom\"\n",
    "  output:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.tsv\"\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_exportTSV_q2.log\"\n",
    "  shell:\n",
    "    \"biom convert -i {input} -o {output} --to-tsv\"\n",
    "\n",
    "rule gen_tax:\n",
    "  input:\n",
    "    sklearn = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" +  PROJ + \"-tax_sklearn.qza\"\n",
    "  output:\n",
    "     table_tax = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\"  + \"taxonomy.tsv\",\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_exportTAXTSV_q2.log\"\n",
    "  params:\n",
    "    directory(SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\")\n",
    "  shell:\n",
    "    \"qiime tools export --input-path {input.sklearn} --output-path {params}\"\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 PHYLOGENETIC TREE \n",
    "##########################################################\n",
    "rule phy_tree:\n",
    "  input:\n",
    "     rep = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-rep-seqs.qza\",\n",
    "  output:\n",
    "    aligned_seqs = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-aligned-rep-seqs.qza\",\n",
    "    aligned_masked = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-masked-aligned-rep-seqs.qza\",\n",
    "    unrooted_tree = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-unrooted-tree.qza\",\n",
    "    rooted_tree = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-rooted-tree.qza\",\n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_phylogeneticTREE_q2.log\"\n",
    "\n",
    "  shell:\n",
    "    \"\"\"qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "        --i-sequences {input.rep} \\\n",
    "        --o-alignment {output.aligned_seqs} \\\n",
    "        --o-masked-alignment {output.aligned_masked} \\\n",
    "        --o-tree {output.unrooted_tree} \\\n",
    "        --o-rooted-tree {output.rooted_tree}\"\"\"\n",
    "\n",
    "\n",
    "##########################################################\n",
    "#                 DIVERSITY METRICS \n",
    "##########################################################\n",
    "rule div_met:\n",
    "  input:\n",
    "     rooted_tree = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + \"tree/\" + PROJ + \"-rooted-tree.qza\",\n",
    "     table = SCRATCH + \"/\" + OUTPUTDIR + \"/asv/\" + PROJ + \"-asv-table.qza\"\n",
    "  output:\n",
    "     output_dir = directory(SCRATCH + \"/qiime2/diversity\")\n",
    "  \n",
    "  log:\n",
    "    SCRATCH + \"/\" + OUTPUTDIR + \"/logs/\" + PROJ + \"_phylogeneticTREE_q2.log\"\n",
    "\n",
    "  shell:\n",
    "    \"\"\"qiime diversity core-metrics-phylogenetic \\\n",
    "        --i-phylogeny {input.rooted_tree} \\\n",
    "        --i-table {input.table} \\\n",
    "        --p-sampling-depth {SAMPLING_DEPTH} \\\n",
    "        --m-metadata-file {METADATA} \\\n",
    "        --output-dir {output.output_dir}\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96d0d32f",
   "metadata": {},
   "source": [
    "# References {.unnumbered}\n",
    "1. https://github.com/shu251/tagseq-qiime2-snakemake/blob/master/Snakefile-asv\n",
    "2. https://forum.qiime2.org/t/qiime2-snakemake-workflow-tutorial-18s-16s-tag-sequencing/11334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa2d146",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
