# Colorectal Cancer Detection
CRC is the second most deadly cancer in the world and could be treated if diagnosed on time. Notably, CRC takes up to ten years to develop from a non-cancerous poly to a cancerous one. This presents a window for early detection of CRC.

Emerging research studies have shown the potential of gut microbiome for detecting CRC. We will utilize datasets from one of those studies to build a CRC classifier. In this project, we’ll walk through **how to build a CRC (Colorectal Carcinoma) classifier for colorectal cancer detection using processed metagenomics data** collected from various colorectal cancer cohorts [@thomas2019metagenomic;@wirbel2019meta]. 

The dataset includes **relative abundance data** of microbial species, categorized by their operation taxonomic units (OTUs). These OTUs organize species across taxonomic ranks such as **kingdom, class, order, phylum, genus, and species**. The dataset draws on samples from multiple published research studies, offering a rich source of analysis. 

:::{.callout-tip}
# Dataset

You can access the processed dataset [here](http://segatalab.cibio.unitn.it/data/Thomas_et_al.html).

:::

## Analysis pipeline
For our analysis, we will follow the steps depicted in the figure below. These steps are taken from refer to analysis pipeline of the same study. 

```{mermaid}
flowchart LR
  A(Data Loading) --> B(Species filtering)
  B --> C(Feature transformation)
  C --> D(Statistical analysis)
  D --> E(Model building)
  F --> E(Model building)
  G --> E(Model building)
  C --> F(Enterotype inference)
  C --> G(PCA)
  E --> H(Generalizability evaluation)
```

* **Data Loading**: The process begins by loading the dataset, which contains relative abundance data and metadata. 

* **Species Filtering**: In this step, microbial species are filtered to retain only those relevant for analysis, improving data quality and reducing noise.
* **Feature Transformation**: The selected features undergo log-transformation to prepare the data for statistical analysis and downstream modeling.
* **Statistical Analysis**: Various statistical techniques are applied to identify significant patterns and associations within the data.
* **Enterotype Inference**: Enterotypes, or gut microbial community types, which were consistently found across different cohorts, are inferred to explore microbiome structure and its potential link to CRC.
* **Principal Component Analysis (PCA)**: PCA reduces data dimensionality, capturing key variance and enhancing model interpretability.
* **Model Building**: This central step integrates insights from statistical analysis, enterotype inference, and PCA to develop the CRC classifier.
* **Generalizability Evaluation**: Finally, the model’s robustness and generalizability are evaluated on datasets from different cohorts.


## References {.unnumbered}